Assets/Mirror/Runtime/Compression.cs

```diff
@@ -0,0 +1,258 @@
+
+            return packed;
+        }
+
+        internal static int FindLargestIndex(Quaternion q)
+        {
+            int index = default;
+            float current = default;
+
+            // check each value to see which one is largest (ignoring +-)
+            for (int i = 0; i < 4; i++)
+            {
+                float next = Mathf.Abs(q[i]);
+                if (next > current)
+                {
+                    index = i;
+                    current = next;
+                }
+            }
+
+            return index;
+        }
+
+        static Vector3 GetSmallerDimensions(int largestIndex, Quaternion value)
+        {
+            float x = value.x;
+            float y = value.y;
+            float z = value.z;
+            float w = value.w;
+
+            float a, b, c;
+            switch (largestIndex)
+            {
+                case 0:
+                    a = y;
+                    b = z;
+                    c = w;
+                    break;
+                case 1:
+                    a = x;
+                    b = z;
+                    c = w;
+                    break;
+                case 2:
+                    a = x;
+                    b = y;
+                    c = w;
+                    break;
+                case 3:
+                    a = x;
+                    b = y;
+                    c = z;
+                    break;
+                default:
+                    return Vector3.zero;
+            }
+
+            return new Vector3(a, b, c);
+        }
+
+
+        /// <summary>
+        /// Used to read a Compressed Quaternion from 4 bytes
+        /// <para>Quaternion is normalized</para>
+        /// </summary>
+        public static Quaternion DecompressQuaternion(uint packed)
+        {
+            // 10 bits
+            const uint mask = 0b11_1111_1111;
+            Quaternion result;
+
+
+            uint a = packed & mask;
+            uint b = (packed >> 10) & mask;
+            uint c = (packed >> 20) & mask;
+            uint largestIndex = (packed >> 30) & mask;
+
+            float x = ScaleFromUInt(a, QuaternionMinValue, QuaternionMaxValue, 0, QuaternionUintRange);
+            float y = ScaleFromUInt(b, QuaternionMinValue, QuaternionMaxValue, 0, QuaternionUintRange);
+            float z = ScaleFromUInt(c, QuaternionMinValue, QuaternionMaxValue, 0, QuaternionUintRange);
+
+            Vector3 small = new Vector3(x, y, z);
+            result = FromSmallerDimensions(largestIndex, small);
+            return result;
+        }
+
+        static Quaternion FromSmallerDimensions(uint largestIndex, Vector3 smallest)
+        {
+            float a = smallest.x;
+            float b = smallest.y;
+            float c = smallest.z;
+
+            float x, y, z, w;
+            float largest = Mathf.Sqrt(1 - a * a - b * b - c * c);
+            switch (largestIndex)
+            {
+                case 0:
+                    x = largest;
```

Comment:
this is too verbose as well.
why not use a simple Vector4 instead of unrolling x,y,z,w?